plugins/scale/scale_title_filter.cpp

```diff
@@ -0,0 +1,339 @@
+        }
+
+        output_scale = output->render->get_target_framebuffer().scale;
+
+        cairo_text_extents_t extents;
+        cairo_font_extents_t font_extents;
+        cairo_select_font_face(cr, "sans-serif", CAIRO_FONT_SLANT_NORMAL,
+            CAIRO_FONT_WEIGHT_BOLD);
+        cairo_set_font_size(cr, font_size * output_scale);
+        cairo_text_extents(cr, title_filter.c_str(), &extents);
+        cairo_font_extents(cr, &font_extents);
+
+        double xpad    = 10.0 * output_scale;
+        double ypad    = 0.2 * (font_extents.ascent + font_extents.descent);
+        unsigned int w = (unsigned int)(extents.width + 2 * xpad);
+        unsigned int h = (unsigned int)(font_extents.ascent +
+            font_extents.descent + 2 * ypad);
+        auto dim = output->get_screen_size();
+        if ((int)w > dim.width)
+        {
+            w = dim.width;
+        }
+
+        if ((int)h > dim.height)
+        {
+            h = dim.height;
+        }
+
+        if ((w > surface_width) || (h > surface_height))
+        {
+            surface_width  = w;
+            surface_height = h;
+            cairo_create_surface();
+        }
+
+        cairo_set_operator(cr, CAIRO_OPERATOR_CLEAR);
+        cairo_paint(cr);
+
+        unsigned int x     = (surface_width - w) / 2;
+        unsigned int y     = (surface_height - h) / 2;
+        unsigned int min_r = (unsigned int)(20 * output_scale);
+        unsigned int r     = h > min_r ? min_r : (h - 2) / 2;
+        const wf::color_t& bg   = bg_color;
+        const wf::color_t& text = text_color;
+
+        cairo_move_to(cr, x + r, y);
+        cairo_line_to(cr, x + w - r, y);
+        cairo_curve_to(cr, x + w, y, x + w, y, x + w, y + r);
+        cairo_line_to(cr, x + w, y + h - r);
+        cairo_curve_to(cr, x + w, y + h, x + w, y + h, x + w - r, y + h);
+        cairo_line_to(cr, x + r, y + h);
+        cairo_curve_to(cr, x, y + h, x, y + h, x, y + h - r);
+        cairo_line_to(cr, x, y + r);
+        cairo_curve_to(cr, x, y, x, y, x + r, y);
+
+        cairo_set_operator(cr, CAIRO_OPERATOR_SOURCE);
+        cairo_set_source_rgba(cr, bg.r, bg.g, bg.b, bg.a);
+        cairo_fill(cr);
+
+        x += xpad;
+        y += ypad + font_extents.ascent;
+        cairo_select_font_face(cr, "sans-serif", CAIRO_FONT_SLANT_NORMAL,
+            CAIRO_FONT_WEIGHT_BOLD);
+        cairo_set_font_size(cr, font_size * output_scale);
+        cairo_move_to(cr, x - extents.x_bearing, y);
+        cairo_set_source_rgba(cr, text.r, text.g, text.b, text.a);
+        cairo_show_text(cr, title_filter.c_str());
+
+        cairo_surface_flush(surface);
+        OpenGL::render_begin();
+        cairo_surface_upload_to_texture(surface, tex);
+        OpenGL::render_end();
+
+        if (!render_active)
+        {
+            output->render->add_effect(&render_hook, wf::OUTPUT_EFFECT_OVERLAY);
+            render_active = true;
+        }
+
+        output->render->damage_whole();
```

Comment:
can't we damage the correct geometry, not the whole output?

Reply:
I did an implementation of this; it still damages the whole overlay surface, since I don't want to save the actual size of the text bubble